doc/R-package/migration_guide.rst

```diff
@@ -0,0 +1,45 @@
+    - Previously, ``xgboost()`` accepted both a ``params`` argument and named arguments under ``...``. Now all training parameters should be passed as named arguments, and all accepted parameters are explicit function arguments with in-package documentation. Some parameters are not allowed as they are determined automatically from the rest of the data, such as the number of classes for multi-classes classification which is determined automatically from 'y'. As well, parameters that have synonyms or which are accepted under different possible arguments (e.g. "eta" and "learning_rate") now accept only their more descriptive form (so "eta" is not accepted, but "learning_rate" is).
+    - Models produced by this function ``xgboost()`` are now returned with a different class "xgboost", which is a subclass of "xgb.Booster" but with more metadata and a ``predict`` method with different defaults.
+    - This function ``xgboost()`` is now meant for interactive usage only. For package developers who wish to incorporate the XGBoost package, it is highly recommended to use ``xgb.train`` instead, which is a lower-level function that closely mimics the same function from the Python package and is meant to be less subject to breaking changes.
+
+- Function ``xgb.train()``:
+    - Previously, ``xgb.train()`` allowed arguments under both a "params" list and as named arguments under ``...``. Now, all training arguments should be passed under ``params``.
+    - In order to make it easier to discover and pass parameters, there is now a function ``xgb.params`` which can generate a list to pass to the ``params`` argument. ``xgb.params`` is simply a function with named arguments that lists everything accepted by ``xgb.train`` and offers in-package documentation for all of the arguments, returning a simple named list.
+    - Arguments that are meant to be consumed by the DMatrix constructor must be passed directly to ``xgb.DMatrix`` instead (e.g. argument for categorical features or for feature names).
+    - Some arguments have been renamed (e.g. previous 'watchlist' is now 'evals', in line with the Python package).
+    - The format of the callbacks to pass to ``xgb.train`` has largely been re-written. See the documentation of ``xgb.Callback`` for details.
+
+- Function ``xgb.DMatrix()``:
+    - This function now accepts 'data.frame' inputs and determines which features are categorical from their types - anything with type 'factor' or 'character' will be considered as categorical. Note that when passing data to the 'predict' method, the 'factor' variables must have the same encoding (i.e. same levels) as XGBoost will not re-encode them for you.
```

Comment:
Factors are not re-encoded at prediction time. But what about characters?

Reply:
They are not recoded either: https://github.com/dmlc/xgboost/blob/88c8d1ae59e48f49adec9c94dcb9e617fee70b13/R-package/R/xgb.Booster.R#L98

Reply:
Unrelated to the PR: this means that single-row predictions with character columns will be completely wrong, right? I feel we should think of fixing this already before the release. 

We would need to 

- turn characters into factors,
- store a named list with names and levels of character/factor columns,
- and maybe a vector of feature names (to allow users to predict on data frames with more/shuffled columns).

`predict()` would need these info to make safe predictions. 

I am aware that we have already discussed on relateed things, but it keeps popping up in my mind.

Reply:
It will have to waiting until support for storing categorical metadata gets implemented in the core library.

Reply:
https://github.com/dmlc/xgboost/issues/11088 will have to be part of the next release. Too much work for 3.0.